code/go/internal/validator/semantic/validate_kibana_no_dangling_object_ids.go

```diff
@@ -4,14 +4,170 @@
 
 package semantic
 
-import "github.com/elastic/package-spec/v2/code/go/internal/errors"
+import (
+	"fmt"
+	"path"
+
+	"github.com/elastic/package-spec/v2/code/go/internal/errors"
+	ve "github.com/elastic/package-spec/v2/code/go/internal/errors"
+	"github.com/elastic/package-spec/v2/code/go/internal/fspath"
+	"github.com/elastic/package-spec/v2/code/go/internal/pkgpath"
+)
+
+type objectReference struct {
+	objectType string
+	objectID   string
+	filePath   string
+}
+
+var exceptionAssets = []string{
+	"index-pattern", // https://github.com/elastic/elastic-package/blob/9d80c6cc282d04f521cd763abd42d529f9679cce/internal/export/transform_filter.go#L14
+}
 
 // ValidateKibanaNoDanglingObjectIDs returns validation errors if there are any
 // dangling references to Kibana objects in any Kibana object files. That is, it
 // returns validation errors if a Kibana object file in the package references another
 // Kibana object with ID i, but no Kibana object file for object ID i is found in the
 // package.
-func ValidateKibanaNoDanglingObjectIDs(pkgRoot string) errors.ValidationErrors {
-	// TODO: will be implemented in follow up PR
-	return nil
+func ValidateKibanaNoDanglingObjectIDs(fsys fspath.FS) errors.ValidationErrors {
+	var errs ve.ValidationErrors
+
+	installedIDs := []objectReference{}
+	referencedIDs := []objectReference{}
+
+	filePaths := path.Join("kibana", "*", "*.json")
+	objectFiles, err := pkgpath.Files(fsys, filePaths)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("error finding Kibana object files: %w", err))
+		return errs
+	}
+	for _, objectFile := range objectFiles {
+		filePath := objectFile.Path()
+
+		currentReference, err := getCurrentObjectReference(objectFile, fsys.Path(filePath))
+		if err != nil {
+			errs = append(errs, fmt.Errorf("unable to create reference from file [%s]: %w", fsys.Path(filePath), err))
+		}
+
+		installedIDs = append(installedIDs, currentReference)
+
+		referencedObjects, err := getReferencesListFromCurrentObject(objectFile, fsys.Path(filePath))
+		if err != nil {
+			errs = append(errs, fmt.Errorf("unable to create referenced objects from file [%s]: %w", fsys.Path(filePath), err))
+			continue
+		}
+
+		referencedIDs = append(referencedIDs, referencedObjects...)
+	}
+
+	if len(referencedIDs) == 0 {
+		return errs
+	}
+
+	for _, reference := range referencedIDs {
+		// look for installed IDs
+		found := false
+		for _, installed := range installedIDs {
+			if reference.objectID != installed.objectID {
+				continue
+			}
+			if reference.objectType != installed.objectType {
+				continue
+			}
+
+			found = true
+		}
```

Comment:
Nit, as the struct seems to be comparable, maybe you can use [`slices.Contains`](https://pkg.go.dev/slices#Contains).
```suggestion
		found := slices.Contains(installedIDs, reference)
```

Reply:
I'll check it, but there is a third parameter `filePath` in the struct that has not the same value between installedIDs and referencedIDs

Reply:
Changed to use `slice.ContainsFunc` to use those specific conditions.

Using `slice.Conditions` make the tests fail.